### PR TITLE
Add workflow to check compatibility

### PR DESCRIPTION
**Pull request by gaiksaya**
**_All the issues linked here confirmed that all plugins have passing `gradle assemble` and `mavenlocal()` as their first repository in build.gradle are closed now: https://github.com/opensearch-project/opensearch-devops/issues/114_**

### Description
This change adds a workflow to check the compatibility with all the plugins in opensearch-project. It adds a comment if any component is incompatible. Example: https://github.com/gaiksaya/OpenSearch/pull/7#issuecomment-1623129382

### Related Issues
https://github.com/opensearch-project/opensearch-devops/issues/114

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: .github/workflows/check-compatibility.yml
Comment (owaiskazi19) [inline .github/workflows/check-compatibility.yml:28]: Can we store the installation_id in secrets?

Comment (gaiksaya) [reply .github/workflows/check-compatibility.yml:28]: Followed what is being used in all other workflows: https://github.com/search?q=repo%3Aopensearch-project%2FOpenSearch%20app_id&type=code

**Comment by gaiksaya** [comment]
Hi @nknize This PR is ready for review. Please let me know if it needs any changes. Thanks!

**Comment by dblock** [comment]
I looked at https://github.com/gaiksaya/OpenSearch/pull/7#issuecomment-1623129382 and it outputs this.

```
> Task :test:external-modules:test-delayed-aggs:publishNebulaPublicationToMavenLocal
> Task :test:external-modules:test-delayed-aggs:publishToMavenLocal
> Task :checkCompatibility
Compatible components: [https://github.com/opensearch-project/common-utils]
```

It tells me what's "incompatible" (I suppose it means it didn't build), but doesn't really give me anything actionable like a log, or a reason. What is one supposed to do when they see an error from this check?

File: .github/workflows/check-compatibility.yml
Comment (dblock) [inline .github/workflows/check-compatibility.yml:5]: This will run on code that's checked in, not on the PR, won't it?

**Comment by gaiksaya** [comment]
> I looked at [gaiksaya#7 (comment)](https://github.com/gaiksaya/OpenSearch/pull/7#issuecomment-1623129382) and it outputs this.
> 
> ```
> > Task :test:external-modules:test-delayed-aggs:publishNebulaPublicationToMavenLocal
> > Task :test:external-modules:test-delayed-aggs:publishToMavenLocal
> > Task :checkCompatibility
> Compatible components: [https://github.com/opensearch-project/common-utils]
> ```
> 
> It tells me what's "incompatible" (I suppose it means it didn't build), but doesn't really give me anything actionable like a log, or a reason. What is one supposed to do when they see an error from this check?

The log or reason can be viewed on the workflow itself: https://github.com/gaiksaya/OpenSearch/actions/runs/5472486309/job/14817298994

Regarding action item, that is next in queue to create GitHub issues in respective component repo once the concerned PR is merged.

File: .github/workflows/check-compatibility.yml
Comment (gaiksaya) [reply .github/workflows/check-compatibility.yml:5]: I believe we want to run on each PR. This will just list the components incompatible or compatible with the change but won't block the PR in anyway unless the gradle task itself fails.

**Comment by bbarani** [comment]
@dblock @gaiksaya Whats the next steps to get this PR merged?

**Review by nknize — APPROVED** [review]
> Whats the next steps to get this PR merged?

I missed this notification sorry for the delay. I'm +1 this PR. It will greatly help me prioritize backports. Thanks @bbarani  and @gaiksaya!

**Comment by gaiksaya** [comment]
Moving forward with this PR. 
Will monitor if this change is working as intended. If it is, the next step as stated above will be creating GH issues in all incompatible components repositories when the pull request breaking things is merged in this repository. 
Thanks!

File: .github/workflows/check-compatibility.yml
Comment (peternied) [reply .github/workflows/check-compatibility.yml:5]: @dblock You are right, this will run the check against whatever is in main, not the incoming pull request.

Note; This what you want from a security perspective otherwise someone could alter the job when making a pull request and have it do Bad Things (tm) to the repository

Comment (peternied) [reply .github/workflows/check-compatibility.yml:5]: Created an issue to track this [1]

- [1] https://github.com/opensearch-project/OpenSearch/issues/9371

Comment (gaiksaya) [reply .github/workflows/check-compatibility.yml:5]: Created pull request to fix this https://github.com/opensearch-project/OpenSearch/pull/9374
Regarding security perspective, we do not integrate GH token with anything related to gradle action. So unless a change happens in core repo workflow, we should be good. https://github.com/opensearch-project/OpenSearch/blob/main/.github/workflows/check-compatibility.yml#L25-L38

**Pull request merged**
